enhancements/microshift/microshift-router-configuration.md

```diff
@@ -0,0 +1,621 @@
+better flexibility)
+
+
+### Allow specific IP addresses
+The following configuration option is proposed:
+```yaml
+router:
+  policy:
+    allow:
+      - <IP address> # Defaults to empty.
+```
+As seen in the previous section the NetworkPolicy resource allows fine grained
+configuration over incoming traffic to the router. NetworkPolicy resources
+work in an "allow" mode, only whats included in them will be allowed to make
+it through these policies.
+
+Applied to he configuration options for the router, users may specify a list
+of allowed IP addresses or CIDRs. This policy allows traffic only from CIDR
+`192.168.122.0/24` to port 443:
+```yaml
+apiVersion: networking.k8s.io/v1
+kind: NetworkPolicy
+metadata:
+  creationTimestamp: "2023-10-31T13:17:42Z"
+  generation: 2
+  name: block-router
+  namespace: openshift-ingress
+  resourceVersion: "5350"
+  uid: e00f5ec7-1557-4285-aa1c-b1b909e53b7a
+spec:
+  ingress:
+  - from:
+    - ipBlock:
+        cidr: 192.168.122.0/24
+    ports:
+    - port: 443
+      protocol: TCP
+  podSelector:
+    matchLabels:
+      ingresscontroller.operator.openshift.io/deployment-ingresscontroller: 
+```
+This will only allow incoming connections to port 443 from network
+`192.168.122.0/24` and deny the rest, even internal traffic.
+
+In order for this to work, and due to MicroShift's CNI, ovnk, we need to make
+further adjustments to services.
+
+The router is exposed through the use of a LoadBalancer service. This service
+will create iptables rules matching on the node IP:
+```bash
+# iptables -s nat -L
+…
+Chain OVN-KUBE-EXTERNALIP (2 references)
+target     prot opt source               destination
+DNAT       tcp  --  anywhere             microshift-dev       tcp dpt:https to:10.43.98.31:443
+DNAT       tcp  --  anywhere             microshift-dev       tcp dpt:http to:10.43.98.31:80
+```
+Traffic towards the router will be DNATted to the node ip, and then reach the
+service. All services default to use `externalTrafficPolicy: Cluster`, which
+will SNAT the client source IP to match that one of the node. This will get
+further NATing as described [here](https://github.com/openshift/microshift/blob/main/docs/contributor/network/ovn_kubernetes_traffic_flows.md#host-to-clusterservice).
+By the time the traffic reaches the NetworkPolicy checks the source IP has
+been replaced with `100.64.0.0/16`, which will not produce a match and traffic
+will get rejected.
+
+To overcome this, all LoadBalancer services in MicroShift that are subject to
+get NetworkPolicy resources to control external traffic must set
+`externalTrafficPolicy: Local`. This setting will not SNAT packets and the
+source IP is preserved until NetworkPolicy checks run.
+
+This setting, however, comes at the expense an even load when having more than
+one node, as traffic will not be sent to another node on the same port. This
+is not an issue in single node, which is MicroShift's use case and design
+principle.
```

Comment:
Does this mean that workloads would have to be configured differently to run on MicroShift and OpenShift?

Reply:
Only LoadBalancer services, doesnt matter who created them. If you want to be able to use ingress NetworkPolicy, you need that setting. This might be a narrow use case, but still.

Reply:
One of the fundamental promises we make is that a user could craft a set of manifests that would work on both MicroShift and OpenShift. We have to retain that ability as we do this new work.

It's not clear to me from your answer if we will be able to do that, but the way a workload would need to be configured would change. Can you clarify that?

Reply:
OK, please capture that detail in the text of the enhancement so we can close this thread.

Reply:
Done